File: 213-house-robber-ii/213-house-robber-ii.java
class Solution {

    public int rob(int[] nums) {
        int n = nums.length;
        if(nums.length == 1)
            return nums[0];
        return Math.max(util(nums, 0, n-2), util(nums,1, n-1));
    }
    
    
    public int util(int[] nums,int start, int end){
        int last = 0, lastSecond = 0;
        for(int i=start;i<=end;i++){
            int curr = Math.max(last, lastSecond+nums[i]);
            lastSecond = last;
            last = curr;
        }
        return last;
    }
}

File: 213-house-robber-ii/213-house-robber-ii.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if(n==1)
            return nums[0];
        if(n==2)
            return max(nums[0], nums[1]);
        return max(util(nums, 0), util(nums, 1));
        
    }
    int util(vector<int>&nums, int start){
        int first = nums[start];
        int second = max(nums[start], nums[start+1]);
        int n = nums.size();
        for(int i=start+3;i<=n-!start;i++){
            int curr = max(nums[i-1]+first, second);
            first = second;
            second = curr;
        }
        return second;
    }
};